daemon/graphdriver/devmapper/deviceset.go

```diff
@@ -841,11 +841,62 @@ func (devices *DeviceSet) createRegisterDevice(hash string) (*devInfo, error) {
 	return info, nil
 }
 
-func (devices *DeviceSet) createRegisterSnapDevice(hash string, baseInfo *devInfo, size uint64) error {
+func (devices *DeviceSet) takeSnapshot(hash string, baseInfo *devInfo, size uint64) error {
+	var doSuspend bool
+	deferredRemove := true
+
 	if err := devices.poolHasFreeSpace(); err != nil {
 		return err
 	}
 
+	if !devices.deferredRemove {
+		deferredRemove = false
+	}
+
+	devinfo, err := devicemapper.GetInfoWithDeferred(baseInfo.Name())
+	if err != nil {
+		return err
+	}
+
+	if devinfo != nil && devinfo.DeferredRemove == 0 {
+		deferredRemove = false
+	}
+
+	if deferredRemove {
+		err = devices._cancelDeferredRemoval(baseInfo)
```

Comment:
How about if we name this new function as cancelDeferredRemoval() and rename the original one to "cancelDeferredRemovalIfNeeded()"